// ngx-paypal/esm2015/lib/models/paypal-models.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
export class PayPalConfig {
    /**
     * @param {?} integrationType
     * @param {?} environment
     * @param {?} config
     */
    constructor(integrationType, environment, config) {
        this.integrationType = integrationType;
        this.environment = environment;
        /**
         * Show 'Pay Now' button config
         */
        this.commit = true;
        Object.assign(this, config);
    }
}
function PayPalConfig_tsickle_Closure_declarations() {
    /**
     * Show 'Pay Now' button config
     * @type {?}
     */
    PayPalConfig.prototype.commit;
    /**
     * Called to create new payment for server side integration
     * @type {?}
     */
    PayPalConfig.prototype.payment;
    /**
     * Called to execute payment for server side integration
     * @type {?}
     */
    PayPalConfig.prototype.onAuthorize;
    /**
     * Client tokens for client side integration
     * @type {?}
     */
    PayPalConfig.prototype.client;
    /**
     * Array of transaction, required for client side integration
     * @type {?}
     */
    PayPalConfig.prototype.transactions;
    /**
     * Called for client side integration when payment is executed
     * @type {?}
     */
    PayPalConfig.prototype.onPaymentComplete;
    /**
     * Button configuration
     * @type {?}
     */
    PayPalConfig.prototype.button;
    /**
     * Paypal funding configuration
     * @type {?}
     */
    PayPalConfig.prototype.funding;
    /**
     * Called when PayPal experiences an error
     * @type {?}
     */
    PayPalConfig.prototype.onError;
    /**
     * Called when user cancels payment
     * @type {?}
     */
    PayPalConfig.prototype.onCancel;
    /**
     * Type of the integration
     * @type {?}
     */
    PayPalConfig.prototype.integrationType;
    /**
     * Environment
     * @type {?}
     */
    PayPalConfig.prototype.environment;
}
/**
 * @record
 */
export function IPayPalPaymentCompleteData() { }
function IPayPalPaymentCompleteData_tsickle_Closure_declarations() {
    /** @type {?} */
    IPayPalPaymentCompleteData.prototype.intent;
    /** @type {?} */
    IPayPalPaymentCompleteData.prototype.orderID;
    /** @type {?} */
    IPayPalPaymentCompleteData.prototype.payerID;
    /** @type {?} */
    IPayPalPaymentCompleteData.prototype.paymentID;
    /** @type {?} */
    IPayPalPaymentCompleteData.prototype.paymentToken;
    /** @type {?} */
    IPayPalPaymentCompleteData.prototype.returnUrl;
}
/**
 * @record
 */
export function IPayPalCancelPayment() { }
function IPayPalCancelPayment_tsickle_Closure_declarations() {
    /** @type {?} */
    IPayPalCancelPayment.prototype.data;
    /** @type {?} */
    IPayPalCancelPayment.prototype.actions;
}
/**
 * @record
 */
export function IPayPalCancelPaymentData() { }
function IPayPalCancelPaymentData_tsickle_Closure_declarations() {
    /** @type {?} */
    IPayPalCancelPaymentData.prototype.billingID;
    /** @type {?} */
    IPayPalCancelPaymentData.prototype.cancelUrl;
    /** @type {?} */
    IPayPalCancelPaymentData.prototype.intent;
    /** @type {?} */
    IPayPalCancelPaymentData.prototype.paymentID;
    /** @type {?} */
    IPayPalCancelPaymentData.prototype.paymentToken;
}
/**
 * @record
 */
export function IPaypalClient() { }
function IPaypalClient_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    IPaypalClient.prototype.sandbox;
    /** @type {?|undefined} */
    IPaypalClient.prototype.production;
}
/**
 * @record
 */
export function IPayPalTransaction() { }
function IPayPalTransaction_tsickle_Closure_declarations() {
    /** @type {?} */
    IPayPalTransaction.prototype.amount;
    /** @type {?|undefined} */
    IPayPalTransaction.prototype.custom;
}
/**
 * @record
 */
export function IPayPalAmount() { }
function IPayPalAmount_tsickle_Closure_declarations() {
    /** @type {?} */
    IPayPalAmount.prototype.total;
    /** @type {?} */
    IPayPalAmount.prototype.currency;
}
/**
 * @record
 */
export function IPayPalButtonStyle() { }
function IPayPalButtonStyle_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    IPayPalButtonStyle.prototype.label;
    /** @type {?|undefined} */
    IPayPalButtonStyle.prototype.size;
    /** @type {?|undefined} */
    IPayPalButtonStyle.prototype.shape;
    /** @type {?|undefined} */
    IPayPalButtonStyle.prototype.color;
    /** @type {?|undefined} */
    IPayPalButtonStyle.prototype.tagline;
    /** @type {?|undefined} */
    IPayPalButtonStyle.prototype.fundingicons;
    /** @type {?|undefined} */
    IPayPalButtonStyle.prototype.branding;
}
/**
 * @record
 */
export function IPayPalFunding() { }
function IPayPalFunding_tsickle_Closure_declarations() {
    /** @type {?} */
    IPayPalFunding.prototype.allowed;
    /** @type {?} */
    IPayPalFunding.prototype.disallowed;
}
